hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java

```diff
@@ -165,7 +165,7 @@ private DataSerializable readInternal(ObjectDataInput in, Class<?> aClass)
             }
 
             ds.readData(in);
-            return ds;
+            return ds.readReplace();
```

Comment:
this has the following consequences:
1. we can use `==` instead of `equals` for `QueryDataType`
2. smaller (but it is only metadata for the query/schema, so not too much?) memory consumption, but the amount of garbage is similar, however after the change it will be more short lived
3. additional virtual method invocation for _each_ DS/IDS serialization

Does 1 and 2 justify 3?

Can we invoke `readReplace` only if some cheap condition is true, eg. flag in the header?

Reply:
A3. We can introduce another interface called `Replaceable` with `readReplace()` method, and check if `ds instanceof Replaceable` before making the call. However, [this Quora answer][1] claims that if a polymorphic method call is usually made against a specific type whose implementation is small enough to be inlined, JIT converts the call into an if-else block by inlining the most common implementation.

A2. We can introduce another interface called `IdentifiedDataSerializer`, which
1. extends `IdentifiedDataSerializable` and `StreamSerializer<? extends DataSerializable>`
2. with `getTypeId()` returning `getClassId()`
3. and `writeData(ObjectDataOutput)`/`readData(ObjectDataInput)` throwing `UnsupportedOperationException`,

and check if `ds instanceof IdentifiedDataSerializer` to decide how to deserialize the object. This approach essentially makes it possible to define a custom serializer that is not affected by `SerializationConfig#allowOverrideDefaultSerializers`. In addition, it will be possible to switch to/from `IdentifiedDataSerializer` in a backward-compatible manner.

[1]: https://www.quora.com/How-many-CPU-instructions-are-typical-for-Java-method-call-overhead/answer/Todd-Lipcon

Reply:
[this](https://shipilev.net/blog/2015/black-magic-method-dispatch/) is a good read, and more up-to-date. However, it does not address our case directly: we have many implementing classes but almost all of them have the same (default) method implementation. Mono/bi-morphic optimisation checks if klass==something not klass!=something

Reply:
`readInternal` method already has a few similar megamoprhic invocations, e.g. `DataSerializableFactory.create`, `ObjectDataInput` method invocations, and not counting creating objects by reflection for DataSerializable. But I would be wary to assume anything regarding the relative cost of additional invocation. We should measure it and ideally analyse the generated code.

Reply:
f81369bd